### ISSUE TITLE
nan loss with poisson objective function

### ISSUE THREAD

**Issue by tboquet** [fchollet/keras#590]
Hello all,

I'm trying to fit a model using the poisson loss function and I get `nan` from the first epoch.
I'm merging 2 models with time dependant sequences/covariates and put a dense layer on the top.

```python
left = Sequential()
left.add(Embedding(len(X_1), 256))
left.add(LSTM(256, 256))
left.add(Dropout(0.3))
left.add(Dense(256, 128))

right = Sequential()
right.add(LSTM(X_1.shape[2], 512, return_sequences=True))
right.add(Dropout(0.3))
right.add(LSTM(512, 256, return_sequences=False))
right.add(Dropout(0.1))
right.add(Dense(256,128))

model = Sequential()
model.add(Merge([left, right], mode='sum'))

model.add(Dense(128, 1))

model.compile(loss='poisson', optimizer='adam')
components = model.fit([X_1, X_2], y, batch_size=256,
                       nb_epoch=20, validation_split=0.1)
```

I've tried to convert my target variable to a 0-1 variable in order to use the binary crossentropy loss and it's working as expected.
I'm not sure of the behavior of the loss function if `y_pred` is 0. The log would be minus infinity and this could be the bug?

I'm not sure how to debug it so if you have some recommandations let me now :).

Thank you!

**Comment by fchollet** [comment]
Right. I fixed this by simply adding a epsilon fuzz factor. Thanks for pointing this out.

**Comment by tboquet** [comment]
I just pulled the latest version and I still have the nan. If `y_pred` is less than 0 it won't work too. I added one line to resolve the problem but I don't know if it's correct from a theoritical point of vue :
```python
y_pred = T.maximum(0, y_pred)
```

**Comment by fchollet** [comment]
If that's not correct, then how would you define Poisson loss in a correct way? It seems to me that the positivity assumption is a reasonable thing to enforce here.

**Comment by tboquet** [comment]
Yep it should be alright and this is the right way to define the poisson loss but I have a weird behavior with this addition:

```python
Train on 90000 samples, validate on 10000 samples
Epoch 0
90000/90000 [==============================] - 184s - loss: 0.7326 - val_loss: 0.8860
Epoch 1
90000/90000 [==============================] - 173s - loss: 0.6638 - val_loss: 0.7031
Epoch 2
90000/90000 [==============================] - 175s - loss: 0.6309 - val_loss: 0.9817
Epoch 3
90000/90000 [==============================] - 176s - loss: 19.1729 - val_loss: 20.7666
Epoch 4
90000/90000 [==============================] - 178s - loss: 19.9556 - val_loss: 20.7666
Epoch 5
90000/90000 [==============================] - 178s - loss: 19.9562 - val_loss: 20.7666
Epoch 6
90000/90000 [==============================] - 175s - loss: 19.9562 - val_loss: 20.7666
Epoch 7
90000/90000 [==============================] - 177s - loss: 19.9562 - val_loss: 20.7666
Epoch 8
90000/90000 [==============================] - 173s - loss: 19.9562 - val_loss: 20.7666
Epoch 9
90000/90000 [==============================] - 176s - loss: 19.9562 - val_loss: 20.7666
```

**Comment by the-moliver** [comment]
Positivity should be enforced but the right place to do it is the activation function for the last layer. The loss function has extreme behavior near 0 so a smooth function like exp or softplus is preferred, while the constraint you implemented is effectively a hard ReLU activation function.

**Comment by tboquet** [comment]
Thank you @the-moliver that's true, let me test. Maybe this loss function should come with a warning (in the documentation) or be declined into several losses?

**Comment by tboquet** [comment]
@the-moliver @fchollet, thank you, the learning is smooth now. Ignore my last comment, I wasn't sure about where to enforce positivity but it's trivial with the activation function.

**Comment by CorySimon** [comment]
Exactly, just to add to this: if you use the `poisson` loss function and `ReLU` loss function to impose positivity, that is inappropriate if your data has zeros. The hacky `epsilon` in the `y_true * log(y_pred+epsilon)` to prevent blow up of the `log` when `y_true` is greater than zero effectively imposes an arbitrary penalty `y_true * log(epsilon)` to every point that your ReLU zeros out. As @the-moliver noted, use a smooth function to impose positivity...

Traditionally ppl model log(\lambda) to impose positivity of lambda. (lambda = mean of Poisson dist'n).

**Comment by Pavlov228** [comment]
Make sure that you don't have negative numbers in your output, since it is impossible to compute logarithm for a negative number. Make sure that you have `relu` activation layer as a last layer in your model to prevent negative outputs.

**Comment by cerlymarco** [comment]
I use this

`def poisson(y_true, y_pred):
    return K.mean(K.maximum(0, y_pred) - y_true * K.log(K.maximum(0, y_pred) + K.epsilon()), axis=-1)`

when compiling

`model.compile(loss=poisson, optimizer = .....)`
